### DISCUSSION TITLE
Use js in main.rs in tauri beta

### DISCUSSION THREAD

**Discussion by timothy-mayweather** [tauri-apps/tauri#3168] [Q&A] [answered]
I have tried to use js in main.rs use webview.eval but it seems it is nolonger supported.

The code:
```rs
#![cfg_attr(
  all(not(debug_assertions), target_os = "windows"),
  windows_subsystem = "windows"
)]

fn main() {
  tauri::Builder::default()
    .setup(|webview| webview.eval("window.location.replace('https://google.com')"))
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```
output:
no method named `eval` found for mutable reference `&mut tauri::App` in the current scope

**Comment by amrbashir** [comment, marked as answer]
the argument in the `setup()` closure is a handle to the tauri app and not the webview window, try this instead.
```rs
use tauri::Manager;

fn main() {
  tauri::Builder::default()
    .setup(|app| {
    	let window = app.get_window("main").unwrap();
        window.eval("window.location.replace('https://google.com')");
        Ok(())
    })
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```

**Reply by timothy-mayweather** [reply]
That is getting me an error.
Error: no method named `get_window` found for mutable reference `&mut tauri::App` in the current scope
My tauri.conf.json
{
  "package": {
    "productName": "title",
    "version": "21.07"
  },
  "build": {
    "distDir": "../dist",
    "devPath": "../dist",
    "beforeDevCommand": "",
    "beforeBuildCommand": ""
  },
  "tauri": {
    "bundle": {
      "active": true,
      "targets": [],
      "identifier": "com.tauri.dev",
      "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ],
      "resources": [],
      "externalBin": [],
      "copyright": "",
      "category": "DeveloperTool",
      "shortDescription": "",
      "longDescription": "",
      "deb": {
        "depends": [],
        "useBootstrapper": false
      },
      "macOS": {
        "frameworks": [],
        "minimumSystemVersion": "",
        "useBootstrapper": false,
        "exceptionDomain": "",
        "signingIdentity": null,
        "entitlements": null
      },
      "windows": {
        "certificateThumbprint": null,
        "digestAlgorithm": "sha256",
        "timestampUrl": ""
      }
    },
    "updater": {
      "active": false
    },
    "allowlist": {
      "all": true,
      "fs": {
        "all":true,
        "writeBinaryFile": true,
        "readDir": true,
      	"readBinaryFile": false,
      	"readTextFile": true,        
      	"writeFile": true 
      },
      "shell": {
        "execute": true
      },
      "dialog": {
        "save": true,
        "open": true
      }
    },
    "windows": [
      {
        "title": "title",
        "width": 800,
        "height": 600,
        "resizable": true,
        "fullscreen": false
      }
    ],
    "security": {
      "csp": "default-src blob: data: filesystem: ws: wss: http: https: tauri: 'unsafe-eval' 'unsafe-inline' 'self' img-src: 'self'"
    }
  }
}

**Reply by amrbashir** [reply]
Sorry about that, you need to import the manager trait
```
use tauri::Manager;
```
I also updated the snippet above.

**Reply by timothy-mayweather** [reply]
Still getting an error. Sorry am not well conversant with rust.
.setup(|app| {
   |  __________________^
11 | |         let window = app.get_window("main").unwrap();
12 | |         window.eval("window.location.replace('https://google.com')");
13 | |     })
   | |_____^ expected enum `Result`, found `()`
   |
   = note:   expected enum `Result<(), Box<(dyn std::error::Error + Send + 'static)>>`
           found unit type `()`

**Reply by amrbashir** [reply]
I updated the snippet again.

**Reply by timothy-mayweather** [reply]
Thanks. Everything is fine now.
